corehq/apps/domain/views.py

```diff
@@ -1975,7 +1975,11 @@ def post(self, request, *args, **kwargs):
                                           name=new_domain.documentation_file_path)
 
             for application in new_domain.full_applications():
-                original_id = application.copied_from._id
+                # Note that application is a build. If the original app has a build then application.copied_from
+                # will be a build and application.copied_from.copy_of will be the original app ID, otherwise
+                # application.copied_from will be the original app. (FB 190587) See also self.published_apps()
```

Comment:
Yeah, [`copied_from`][1] is populated using the last item in `copy_history`, and `copy_history` is populated in [domain/models.py:649][2] and [758][3], both of which get called by (among other things) [`save_snapshot()`][4] a few lines up in domain/views.py. I found `copy_history` a little confusing, because it doesn't seem to point to the previous version -- the copy_history of the second version of my test app only had one item in it -- it points to an Application instance that includes attachments of the contents of `commcare.ccz` and has `copy_of` set to the app ID (which happens [when it's built][6]).

I couldn't see where `copy_history` is set for an app that hasn't been built yet -- maybe because it's populated from `original_doc` [when it's wrapped][5], but that looks like legacy, so probably not.

domain/views.py does look ripe for a little refactoring, maybe delegating chunks of longer views to util functions, and renaming `application` variables to `original_app`, `the_kind_of_build_with_copy_history` (whatever we call that thing) and `the_kind_of_build_with_ccz_contents` (whatever we call that thing too). I'm sure someone who knows this stuff can help me with the names. :grin:

  [1]: https://github.com/dimagi/commcare-hq/blob/a0bf223a6f42c72c8e5bb09f2c900bcee0972057/corehq/apps/appstore/models.py#L16-L16
  [2]: https://github.com/dimagi/commcare-hq/blob/e4045bc13636f6d980318a8e83592ad3ced2480c/corehq/apps/domain/models.py#L649-L649
  [3]: https://github.com/dimagi/commcare-hq/blob/e4045bc13636f6d980318a8e83592ad3ced2480c/corehq/apps/domain/models.py#L758-L758
  [4]: https://github.com/dimagi/commcare-hq/blob/e4045bc13636f6d980318a8e83592ad3ced2480c/corehq/apps/domain/models.py#L796-L796
  [5]: https://github.com/dimagi/commcare-hq/blob/0b8219bb72370aa70bb8f352eeb496e5e8ff788f/corehq/apps/app_manager/models.py#L3916-L3916
  [6]: https://github.com/dimagi/commcare-hq/blob/0b8219bb72370aa70bb8f352eeb496e5e8ff788f/corehq/apps/app_manager/models.py#L3706-L3706